### DISCUSSION TITLE
How to use faiss.index_cpu_to_all_gpus() for faiss.IndexBinaryFlat?

### DISCUSSION THREAD

**Discussion by anoubhav** [facebookresearch/faiss#3646] [Q&A]
I want to use multiple GPUs while using the binary flat index. 
When I run `faiss.index_cpu_to_all_gpus(faiss.IndexBinaryFlat(d))`, I get the following error:

```
TypeError: Wrong number or type of arguments for overloaded function 'index_cpu_to_gpu_multiple'.
  Possible C/C++ prototypes are:
    faiss::gpu::index_cpu_to_gpu_multiple(std::vector< faiss::gpu::GpuResourcesProvider * > &,std::vector< int > &,faiss::Index const *,faiss::gpu::GpuMultipleClonerOptions const *)
    faiss::gpu::index_cpu_to_gpu_multiple(std::vector< faiss::gpu::GpuResourcesProvider * > &,std::vector< int > &,faiss::Index const *)
```

I don't get this error when using L2 flat index. `faiss.index_cpu_to_all_gpus(faiss.IndexFlatL2(d))`

**Comment by mdouze** [comment]
faiss.index_cpu_to_all_gpus is only for float indexes. 
For binary indexes (whether CPU or GPU), the construction should be done manually. 
See here for an example on how to do that: 
https://github.com/facebookresearch/faiss/blob/7492d233541acdab1fc8c7e7a6bb02b8ad535e1f/tests/test_index_binary.py#L341

**Comment by anoubhav** [comment]
Just to clarify: "construction should be done manually" means that I cannot train a binary index with a GPU compatible binary clustering index?

https://gist.github.com/mdouze/dd11f1ebd2f1c2f3bcd74beee303e513 
In the above link, you shared how to use a GPU-compatible faiss.IndexFlatL2 to build a binary IVF index.

Is it possible to use GPU-optimized `faiss.IndexBinaryFlat` to build a binary IVF index?
Thanks,

**Comment by anoubhav** [comment]
Is it possible to perform a search using `faiss.IndexBinaryFlat` on multiple GPUs?

Using `faiss.GpuIndexBinaryFlat` I am able to do the search on a single GPU. But not on more than 1 GPUs due to the above error when trying the binary flat index in `faiss.index_cpu_to_all_gpus`
